contributors/design-proposals/admission_control_extension.md

```diff
@@ -0,0 +1,265 @@
+party resources or API extension servers), these APIs have many of the same security and policy needs that
+the core resources do, but must compile in their own subsets of admission.
+
+Further, distributions of Kubernetes like OpenShift that wish to offer complete solutions (such as OpenShift's
+multi-tenancy model) have no mechanism for running on top of Kubernetes without recompilation of the core or
+for extending the core with additional policy. This prevents the formation of an open ecosystem for tools
+*around* Kubernetes, forcing all changes to policy to go through the Kubernetes codebase review gate (when
+such review is unnecessary or disruptive to Kubernetes itself).
+
+
+## Design
+
+It should be possible to perform holistic policy enforcement in Kubernetes without the recompilation of the
+core project as plugins that can be added and removed to a stock Kubernetes release. That extension
+of admission control should leverage similar mechanisms to our existing controller frameworks where possible
+and otherwise be performant and reliable.
+
+
+### Requirements
+
+1.  Easy Initialization
+    Privileged components should be able to easily participate in the **initialization** of a new object.
+2.  Synchronous Validation
+    Synchronous rejection of initialized objects or mutations must be possible outside of the kube-apiserver binary
+3.  Backwards Compatible
+    Existing API clients must see no change in behavior to external admission other than increased latency
+4.  Easy Installation
+    Administrators should be able to easily write a new admission plugin and deploy it in the cluster
+5.  Performant
+    External admission must not significantly regress performance in large and dense clusters
+6.  Reliable
+    External admission should be capable of being "production-grade" for deployment in an extremely large and dense cluster
+7.  Internally Consistent
+    Developing an admission controller should reuse as much infrastructure and tools as possible from building custom controllers so as to reduce the cost of extension.
+
+
+### Specification
+
+Based on observation of the actual admission control implementations the majority of mutation
+occurs as part of creation, and a large chunk of the remaining controllers are for side-effect free
+validation of creation and updates. Therefore we propose the following changes to Kubernetes:
+
+1.  Allow some controllers to act as "initializers" - watching the API and mutating the object before it is visible to normal clients.
+    This would reuse the majority of the infrastructure in place for controllers. Because creation is
+    one-way, the object can be "revealed" to regular clients once a set list of initializers is consumed. These
+    controllers could run on the cluster as pods.
+2.  Add a generic **external admission webhook** controller that is non-mutating (thus parallelizable)
```

Comment:
while I like the idea of being able to parallelize these, I think you'll still want some mechanism in there so that an admin can require some set be run in serial before all others - e.g. auth type of checks so that people don't even get a chance to invoke (and potential get exposure to the inner workings of things) that they shouldn't even know exist. A security concern.